### PR TITLE
RBAC: Optimize permissions caching

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What this PR does / why we need it**:

This PR optimizes permissions cache usage:

- Cache result for the whole `getCachedUserPermissions()` call. So if user's permissions found in cache, return immediately.
- If not, use separate cache for basic roles, teams and managed permissions. This will help to reduce `getUserPermissions()` time for other users by sharing teams and basic roles permissions cache.
- Clear user's teams permissions when clearing user's permissions cache

This helps us to reach target RPS number under the load by optimizing request hot path.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/services/accesscontrol/acimpl/service.go [hunk 1]
```diff
@@ -373,8 +388,15 @@ func (s *Service) getCachedTeamsPermissions(ctx context.Context, user identity.R
...
-	s.cache.Delete(accesscontrol.GetPermissionCacheKey(user))
+	s.cache.Delete(accesscontrol.GetUserPermissionCacheKey(user))
 	s.cache.Delete(accesscontrol.GetUserDirectPermissionCacheKey(user))
+
+	// Clear user's teams cache
+	teams := user.GetTeams()
+	orgID := user.GetOrgID()
+	for _, teamID := range teams {
+		s.cache.Delete(accesscontrol.GetTeamPermissionCacheKey(teamID, orgID))
+	}
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/acimpl/service.go:399]: From what I can see, we only call `ClearUserPermissionCache` when:
* a new dash/folder/data source/service account/team is created;
* user's basic org role is updated.

In none of these scenarios the perms granted to the existing teams change. So I think it's fine to not clear team perm cache here. But I do think we should be better at clearing caches in general (eg, if a team is granted a new permission).

**Review by IevaVasiljeva — APPROVED** [review]
LGTM overall. I left one comment about cache clearing logic. I also noticed that we don't cache permissions for teams that have none. Eg, if team has no permissions assigned to it, we'll keep trying to fetch them, as we never set cache for this team (cache gets set when we loop over team with perms [here](https://github.com/grafana/grafana/pull/92673/files#diff-6ffbe8013330badb7e88e3f022a1cc5f9e6048363e36e2e07d5b73a652e679c3R380-R384)). I don't think it's a big deal though, as most teams should have some perms assigned to them.

Apart from that looks good and works as expected. I tested the following scenarios:
* viewer user and admin user with no teams;
* viewer user and admin user with the same set of teams;
* user who is viewer in one org and admin in another one.

File: pkg/services/accesscontrol/acimpl/service.go (hunk 1 above)
Comment (alexanderzobnin) [reply pkg/services/accesscontrol/acimpl/service.go:399]: Yeah, I also thought about this. It basically has nothing to do with teams permissions. I'll remove this.

File: pkg/services/accesscontrol/acimpl/service.go [hunk 2]
```diff
@@ -241,32 +242,39 @@ func (s *Service) getCachedUserPermissions(ctx context.Context, user identity.Re
...
 
 	return permissions, nil
 }
 
-func (s *Service) getCachedBasicRolesPermissions(ctx context.Context, user identity.Requester, options accesscontrol.Options, permissions []accesscontrol.Permission) ([]accesscontrol.Permission, error) {
+func (s *Service) getCachedBasicRolesPermissions(ctx context.Context, user identity.Requester, options accesscontrol.Options) ([]accesscontrol.Permission, error) {
 	ctx, span := tracer.Start(ctx, "accesscontrol.acimpl.getCachedBasicRolesPermissions")
 	defer span.End()
 
+	permissions := make([]accesscontrol.Permission, 0)
```
Comment (gamab) [inline pkg/services/accesscontrol/acimpl/service.go:277]: Slightly more efficient to pre-allocate the slice with a reasonable number.
```suggestion
	permissions := make([]accesscontrol.Permission, 0, 50)
```

File: pkg/services/accesscontrol/acimpl/service.go [hunk 3]
```diff
@@ -351,6 +363,9 @@ func (s *Service) getCachedTeamsPermissions(ctx context.Context, user identity.R
 				miss = append(miss, teamID)
 			}
 		}
+	} else {
+		// reload cache and fetch all teams permissions
+		miss = teams
```
Comment (gamab) [inline pkg/services/accesscontrol/acimpl/service.go:368]: :+1:

**Review by gamab — APPROVED** [review]
LGTM

**Pull request merged**